### ISSUE TITLE
Orchard OutputCache and OutputCache HttpModule

### ISSUE THREAD

**Issue by jtkech** [OrchardCMS/Orchard#3]
Following the recent discussions on OutputCache in the Committee videos, about public caching, iis kernel cache, and the potential issue with the capture stream...

So, especially if Orchard OutputCache remains not compatible with some content types (but not only), and because the HttpModule is no longer removed in web.config (I didn't notice that), my idea here is to be able, for a specific controller action, to prevent Orchard caching and use the OutputCache HttpModule in a normal way

This, not only with ignoredUrls if many different files can be served by a controller, but maybe by only testing if an [outputCache] attribute is used, even with Duration > 0 and NoStore = false. Of course, for these specific content types we will not have the same Orchard caching rules, but the ability to use e.g all the OutputCacheLocation options...

I've done some tests, with an image file (23 kb) served by a controller, to see server side caching response times

- __No caching__ at all: Response time __12 ms__

- __Orchard caching__, maxAge = 0 => no proxies / client caching: Response time __6 ms__

- __HttpModule__, [OutputCache] attribute with Duration > 0, and OutputCacheLocation.Server => Cache-Control: no-cache, but caching is done on server: Response time __3 ms__

Best

**Comment by DanielStolt** [comment]
I still don't believe there actually is any issue with the filter capture stream. I intend to test it with IIS Kernel Cache to make sure, but I haven't gotten around to it yet. Is that among the scenarios you have tested?

**Comment by jtkech** [comment]
Right now, on my local machine I only have iis express, I've tried this command: "netsh http show cachestate url=..." but there is no cache entry. I think we can't use kernel cache with iis express

And all my sites are on Azure WebSites where I don't have the right access to do net shell commands, I've tried it with the kudu console. That said, I can't repro the issue about duplicated responses described by Sebastien, but I'm not sure that the kernel cache is used in these tests...

For infos, to use the kernel cache one condition is to have "HttpCacheability.Public", I've checked it in this source file at the IsKernelCacheable() method:
https://github.com/Microsoft/referencesource/blob/master/System.Web/HttpCachePolicy.cs

So, the kernel cache can be considered as a public proxy but there are some interactions with the .net framework running on the server...

But, in my previous post, it's not about kernel cache because with Orchard caching I use maxAge = 0 => the default "HttpCacheability.Private" => Cache-Control: private (no caching on proxies, and therefore no kernel cache). And when I only use the HttpModule it's with OutputCacheLocation.Server => "HttpCacheability.Server" => Cache-Control: no-cache (but caching is done on server)

More details here
https://msdn.microsoft.com/fr-fr/library/system.web.ui.outputcachelocation%28v=vs.110%29.aspx
https://msdn.microsoft.com/fr-fr/library/system.web.httpcacheability(v=vs.110).aspx

So, in my previous post I only talk about the ability to use the HttpModule in a regular way. This can be useful if Orchard caching remains not compatible with some content types, but not only. For example, if you don't need Orchard caching rules for some file results, it would be good to only use the HttpModule that seems very efficient. Maybe in Orchard OutputCache, we can only test if an [outputCache] attribute is used, even with Duration > 0...

Finally, I'm very interesting to know if you can repro the duplicated responses issue. This, because if Orchard caching can be compatible with all content types, it would be ok for me, even if we can't use only the HttpModule for some file results...

If I can get a machine with IIS, I'll do more testing

Best

**Comment by DanielStolt** [comment]
I've spent some time working on this tonight.

First of all, I have tested all scenarios I can think of and I have not found any issues with the filter stream approach, neither duplicated content nor any other issues.

Secondly, I have spent several hours trying to get IIS kernel cache to work, but it simply won't, neither in 1.9-int nor in feature/outputcache. The only way I can get it to work is to enable it in Web.config using a `<cachingPolicy>` element, but this is not a good approach, as it is meant for files, assumes file notification-based invalidation, ignores query strings, etc. So, either way, right now IIS kernel cache is a no-go, with our without my changes.

Thirdly, since you mentioned that OutputCacheModule is no longer removed in Web.config, I looked into that too. It was re-added by Sebastien last year because he found it was required in order for IIS kernel cache to work. However, I did some checking and found that with OutputCacheModule we are actually caching everything double - once in Orchard.OutputCache and once in ASP.NET output cache. And since IIS kernel cache does not work anyway, with or without OutputCacheModule, I have once again removed it.

Given the above results, I have now merged feature/outputcache into 1.9-int.

In other words, Orchard.OutputCache should now work fine with all result types, properly ignore child actions, etc. And hence, there should be no need for you to use the ASP.NET OutputCacheModule.

However, if despite all this you should still want to use ASP.NET OutputCacheModule, you can still do that by:

1. Commenting out the `<remove>` element in Web.config to reactivate ASP.NET OutputCacheModule.
2. Disabling Orchard.OutputCache for the route in Cache Settings (set Duration to 0).
3. Decorating your controller and/or action with the `[OutputCache]` attribute (this will activate ASP.NET output caching for the request).

**Comment by jtkech** [comment]
Wow, if compatible with all result types, and by default the OutputCacheModule is removed, all is perfect for me

In this context, I don't need OutputCacheModule, but, just for infos, I've tried your solution. If I disable OutputCache for the route in Cache Settings => ResponseIsCacheable() return false => then the code set in the Response Headers no-cache, no-store and maxAge = 0, then there is no caching at all, even if I use an [OutputCache] attribute with Duration > 0

Anyway, great work Daniel, thanks

**Comment by DanielStolt** [comment]
Ah, yes, I didn't consider that OutputCacheFilter would set no-cache, preventing ASP.NET OutputCacheModule from doing any caching... anyway, glad it worked out.
